lib/gemstash/S3.rb

```diff
@@ -0,0 +1,247 @@
+      save_properties(properties.merge(props || {}))
+      self
+    end
+
+    def property?(*keys)
+      keys.inject(node: properties, result: true) do |memo, key|
+        if memo[:result]
+          memo[:result] = memo[:node].is_a?(Hash) && memo[:node].include?(key)
+          memo[:node] = memo[:node][key] if memo[:result]
+        end
+
+        memo
+      end[:result]
+    end
+
+    def delete(key)
+      return self unless exist?(key)
+
+      begin
+        @S3resource.object(content_filename(key)).delete
+      rescue StandardError => e
+        log_error "Failed to delete stored content at #{content_filename(key)}", e, level: :warn
+      end
+
+      begin
+        @S3resource.object(properties_filename).delete unless content?
+      rescue StandardError => e
+        log_error "Failed to delete stored properties at #{properties_filename}", e, level: :warn
+      end
+
+      self
+    ensure
+      reset
+    end
+
+
+    private
+    def content?
+      return false unless @S3resource.object(@folder).exists?
+
+      entries = @S3resource.objects(prefix: (@folder)).collect().reject { |object| object.content_length == 0 || object.key == "properties.yaml" }
+      !entries.empty?
+    end
+
+    def load_properties(force = false)
+      return if @properties && !force
+      return unless @S3resource.object(properties_filename).exists?
+
+      properties_file = @S3resource.object(properties_filename).get.body
+      @properties = YAML.load(properties_file) || {}
+      check_resource_version
+    end
+
+    def save_properties(props)
+      props ||= {}
+      props = { gemstash_resource_version: Gemstash::S3Resource::VERSION }.merge(props)
+      store(properties_filename, props.to_yaml)
+      @properties = props
+    end
+
+    def load(key)
+      raise "Resource #{@name} has no #{key.inspect} content to load" unless exist?(key)
+      load_properties # Ensures storage version is checked
+      @content ||= {}
+      @content[key] = read_file(content_filename(key))
+    end
+
+    def check_resource_version
+      version = @properties[:gemstash_resource_version]
+      return if version <= Gemstash::S3Resource::VERSION
+
+      reset
+      raise Gemstash::S3Resource::VersionTooNew.new(name, folder, version)
+    end
+
+    def sanitize(name)
+      name.gsub(/[^a-zA-Z0-9_]/, "_")
+    end
+
+    def save_content(key,content)
+      store(content_filename(key), content)
+      @content ||= {}
+      @content[key] = content
+    end
+
+    def store(filename,content)
+      save_file(filename) { content }
+    end
+
+    def content_filename(key)
+      name = sanitize(key.to_s)
+      raise "Invalid content key #{key.inspect}" if name.empty?
+
+      File.join(@folder, name)
+    end
+
+    def read_file(filename)
+      @S3resource.object(filename).get().body.read
+    end
+
+    def properties_filename
+      File.join(@folder, "properties.yaml")
+    end
+
+    def save_file(filename)
+      content = yield
```

Comment:
why we need to yield the content and to pass it as an argument?

Reply:
In hindsight, this method yield is certainly a bit unnecessary. I leave it as is because it was already implemented in the previous storage class so I didn't want to cause unexpected errors. At some point, I will refactor the entire storage class to make it simpler, and more straightforward.